# Wragle Twitter data for We Rate Dogs twitter account
Going to use tweepy package to do this. The below code initializes 

In [1]:
import tweepy

# import twitter keys and tokens required for initializing api
from twitter_api_keys import consumer_key, consumer_secret, access_token, access_secret

# the above line will not work for you if you just clone the repo.
# To make it work, set the above variables to your specific values that you get when you
# create your twitter app.


auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth)

## Gather
Import necessary libraries and read the json twitter data, as well as the 'twitter-archive-enhanced.csv'.
This csv was provided as part of the project and was downloaded manually from the link: https://d17h27t6h515a5.cloudfront.net/topher/2017/August/59a4e958_twitter-archive-enhanced/twitter-archive-enhanced.csv

In [2]:
import pandas as pd

In [3]:
tweet_archive_df = pd.read_csv('twitter-archive-enhanced.csv')
tweet_archive_df.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [4]:
tweet_archive_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

#### Checking the json data for the first tweet id just to check


In [5]:
check_tweet = api.get_status(tweet_archive_df.tweet_id[0], tweet_mode='extended')

In [13]:
# Checking all available methods for the status object
dir(check_tweet)

['__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 '_api',
 '_json',
 'author',
 'contributors',
 'coordinates',
 'created_at',
 'destroy',
 'display_text_range',
 'entities',
 'extended_entities',
 'favorite',
 'favorite_count',
 'favorited',
 'full_text',
 'geo',
 'id',
 'id_str',
 'in_reply_to_screen_name',
 'in_reply_to_status_id',
 'in_reply_to_status_id_str',
 'in_reply_to_user_id',
 'in_reply_to_user_id_str',
 'is_quote_status',
 'lang',
 'parse',
 'parse_list',
 'place',
 'possibly_sensitive',
 'possibly_sensitive_appealable',
 'retweet',
 'retweet_count',
 'retweeted',
 'retweets',
 'source',
 'source_url',
 'truncated',
 'user']

In [14]:
# using the full_text attribute of the status object
check_tweet.full_text

"This is Phineas. He's a mystical boy. Only ever appears in the hole of a donut. 13/10 https://t.co/MgUWQ76dJU"

In [15]:
import json

In [17]:
type(check_tweet)

tweepy.models.Status

In [31]:
str(check_tweet)

'Status(_api=<tweepy.api.API object at 0x0000023B5EA2EEF0>, _json={\'created_at\': \'Tue Aug 01 16:23:56 +0000 2017\', \'id\': 892420643555336193, \'id_str\': \'892420643555336193\', \'full_text\': "This is Phineas. He\'s a mystical boy. Only ever appears in the hole of a donut. 13/10 https://t.co/MgUWQ76dJU", \'truncated\': False, \'display_text_range\': [0, 85], \'entities\': {\'hashtags\': [], \'symbols\': [], \'user_mentions\': [], \'urls\': [], \'media\': [{\'id\': 892420639486877696, \'id_str\': \'892420639486877696\', \'indices\': [86, 109], \'media_url\': \'http://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg\', \'media_url_https\': \'https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg\', \'url\': \'https://t.co/MgUWQ76dJU\', \'display_url\': \'pic.twitter.com/MgUWQ76dJU\', \'expanded_url\': \'https://twitter.com/dog_rates/status/892420643555336193/photo/1\', \'type\': \'photo\', \'sizes\': {\'thumb\': {\'w\': 150, \'h\': 150, \'resize\': \'crop\'}, \'medium\': {\'w\': 540, \'h\': 528, \

In [37]:
tweet_archive_df.tweet_id.head().values

array([892420643555336193, 892177421306343426, 891815181378084864,
       891689557279858688, 891327558926688256], dtype=int64)

In [50]:
json_str_list = []
exception_list = []
import sys

## Testing the code to download the twitter json data on sample

In [61]:
with open('tweet-json-test.txt','w') as json_file:
    for tweetid in tweet_archive_df.tweet_id.tail().values:
        print('getting data for: ',tweetid)
        try:
            status = api.get_status(tweetid, tweet_mode='extended')
            json_str_list.append( json.dumps(status._json) )
        except:
            exception_list.append(tweetid)
            print('exception_raised for: ',tweetid)
            print(sys.exc_info()[0])
    json.dump(json_str_list,json_file)

    

getting data for:  666049248165822465
getting data for:  666044226329800704
getting data for:  666033412701032449
getting data for:  666029285002620928
getting data for:  666020888022790149


### Test

In [62]:
with open('tweet-json-test.txt','r') as json_file:
    data = json.load(json_file)
print(len(data))
# json.load

10


## Gathering all twitter data for the given tweet ids
The above programmatic download code for the twitter json data will now be used to download all 2356 tweet ids.
<br> The tweet ids that have been deleted will be catched in the exception list, which I will check later

In [53]:
len(tweet_archive_df)

2356

In [63]:
with open('tweet-json.txt','w') as json_file:
    for tweetid in tweet_archive_df.tweet_id.values:
        print('getting data for: ',tweetid)
        try:
            status = api.get_status(tweetid, tweet_mode='extended')
            json_str_list.append( json.dumps(status._json) )
        except:
            exception_list.append(tweetid)
            print('exception_raised for: ',tweetid)
            print(sys.exc_info()[0])
    json.dump(json_str_list,json_file)

    

getting data for:  892420643555336193
getting data for:  892177421306343426
getting data for:  891815181378084864
getting data for:  891689557279858688
getting data for:  891327558926688256
getting data for:  891087950875897856
getting data for:  890971913173991426
getting data for:  890729181411237888
getting data for:  890609185150312448
getting data for:  890240255349198849
getting data for:  890006608113172480
getting data for:  889880896479866881
getting data for:  889665388333682689
getting data for:  889638837579907072
getting data for:  889531135344209921
getting data for:  889278841981685760
getting data for:  888917238123831296
getting data for:  888804989199671297
getting data for:  888554962724278272
getting data for:  888202515573088257
exception_raised for:  888202515573088257
<class 'tweepy.error.TweepError'>
getting data for:  888078434458587136
getting data for:  887705289381826560
getting data for:  887517139158093824
getting data for:  887473957103951883
getting data

getting data for:  854010172552949760
getting data for:  853760880890318849
getting data for:  853639147608842240
getting data for:  853299958564483072
getting data for:  852936405516943360
getting data for:  852912242202992640
getting data for:  852672615818899456
getting data for:  852553447878664193
getting data for:  852311364735569921
getting data for:  852226086759018497
getting data for:  852189679701164033
getting data for:  851953902622658560
getting data for:  851861385021730816
getting data for:  851591660324737024
getting data for:  851464819735769094
getting data for:  851224888060895234
getting data for:  850753642995093505
getting data for:  850380195714523136
getting data for:  850333567704068097
getting data for:  850145622816686080
getting data for:  850019790995546112
getting data for:  849776966551130114
getting data for:  849668094696017920
getting data for:  849412302885593088
getting data for:  849336543269576704
getting data for:  849051919805034497
getting data

getting data for:  823719002937630720
getting data for:  823699002998870016
getting data for:  823581115634085888
getting data for:  823333489516937216
getting data for:  823322678127919110
getting data for:  823269594223824897
getting data for:  822975315408461824
getting data for:  822872901745569793
getting data for:  822859134160621569
getting data for:  822647212903690241
getting data for:  822610361945911296
getting data for:  822489057087389700
getting data for:  822462944365645825
getting data for:  822244816520155136
getting data for:  822163064745328640
getting data for:  821886076407029760
getting data for:  821813639212650496
getting data for:  821765923262631936
getting data for:  821522889702862852
getting data for:  821421320206483457
getting data for:  821407182352777218
getting data for:  821153421864615936
getting data for:  821149554670182400
getting data for:  821107785811234820
getting data for:  821044531881721856
getting data for:  820837357901512704
getting data

getting data for:  796177847564038144
getting data for:  796149749086875649
getting data for:  796125600683540480
getting data for:  796116448414461957
getting data for:  796080075804475393
getting data for:  796031486298386433
getting data for:  795464331001561088
getting data for:  795400264262053889
getting data for:  795076730285391872
getting data for:  794983741416415232
getting data for:  794926597468000259
getting data for:  794355576146903043
getting data for:  794332329137291264
getting data for:  794205286408003585
getting data for:  793962221541933056
getting data for:  793845145112371200
getting data for:  793614319594401792
getting data for:  793601777308463104
getting data for:  793500921481273345
getting data for:  793286476301799424
getting data for:  793271401113350145
getting data for:  793256262322548741
getting data for:  793241302385262592
getting data for:  793226087023144960
getting data for:  793210959003287553
getting data for:  793195938047070209
getting data

getting data for:  768855141948723200
getting data for:  768609597686943744
getting data for:  768596291618299904
getting data for:  768554158521745409
getting data for:  768473857036525572
getting data for:  768193404517830656
getting data for:  767884188863397888
getting data for:  767754930266464257
getting data for:  767500508068192258
getting data for:  767191397493538821
getting data for:  767122157629476866
getting data for:  766864461642756096
getting data for:  766793450729734144
getting data for:  766714921925144576
getting data for:  766693177336135680
getting data for:  766423258543644672
getting data for:  766313316352462849
getting data for:  766078092750233600
getting data for:  766069199026450432
getting data for:  766008592277377025
getting data for:  765719909049503744
getting data for:  765669560888528897
getting data for:  765395769549590528
getting data for:  765371061932261376
getting data for:  765222098633691136
getting data for:  764857477905154048
getting data

getting data for:  743609206067040256
getting data for:  743595368194129920
getting data for:  743545585370791937
getting data for:  743510151680958465
getting data for:  743253157753532416
getting data for:  743222593470234624
getting data for:  743210557239623680
getting data for:  742534281772302336
getting data for:  742528092657332225
getting data for:  742465774154047488
getting data for:  742423170473463808
getting data for:  742385895052087300
getting data for:  742161199639494656
getting data for:  742150209887731712
getting data for:  741793263812808706
getting data for:  741743634094141440
getting data for:  741438259667034112
getting data for:  741303864243200000
getting data for:  741099773336379392
getting data for:  741067306818797568
getting data for:  740995100998766593
getting data for:  740711788199743490
getting data for:  740699697422163968
getting data for:  740676976021798912
getting data for:  740373189193256964
getting data for:  740365076218183684
getting data

getting data for:  710269109699739648
getting data for:  710153181850935296
getting data for:  710140971284037632
getting data for:  710117014656950272
getting data for:  709918798883774466
getting data for:  709901256215666688
getting data for:  709852847387627521
getting data for:  709566166965075968
getting data for:  709556954897764353
getting data for:  709519240576036864
getting data for:  709449600415961088
getting data for:  709409458133323776
getting data for:  709225125749587968
getting data for:  709207347839836162
getting data for:  709198395643068416
getting data for:  709179584944730112
getting data for:  709158332880297985
getting data for:  709042156699303936
getting data for:  708853462201716736
getting data for:  708845821941387268
getting data for:  708834316713893888
getting data for:  708810915978854401
getting data for:  708738143638450176
getting data for:  708711088997666817
getting data for:  708479650088034305
getting data for:  708469915515297792
getting data

getting data for:  693629975228977152
getting data for:  693622659251335168
getting data for:  693590843962331137
getting data for:  693582294167244802
getting data for:  693486665285931008
getting data for:  693280720173801472
getting data for:  693267061318012928
getting data for:  693262851218264065
exception_raised for:  693262851218264065
<class 'tweepy.error.TweepError'>
getting data for:  693231807727280129
getting data for:  693155686491000832
getting data for:  693109034023534592
getting data for:  693095443459342336
getting data for:  692919143163629568
getting data for:  692905862751522816
getting data for:  692901601640583168
getting data for:  692894228850999298
getting data for:  692828166163931137
getting data for:  692752401762250755
getting data for:  692568918515392513
getting data for:  692535307825213440
getting data for:  692530551048294401
getting data for:  692423280028966913
getting data for:  692417313023332352
getting data for:  692187005137076224
getting data

getting data for:  681339448655802368
getting data for:  681320187870711809
getting data for:  681302363064414209
getting data for:  681297372102656000
getting data for:  681281657291280384
getting data for:  681261549936340994
getting data for:  681242418453299201
getting data for:  681231109724700672
getting data for:  681193455364796417
getting data for:  680970795137544192
getting data for:  680959110691590145
getting data for:  680940246314430465
getting data for:  680934982542561280
getting data for:  680913438424612864
getting data for:  680889648562991104
getting data for:  680836378243002368
getting data for:  680805554198020098
getting data for:  680801747103793152
getting data for:  680798457301471234
getting data for:  680609293079592961
getting data for:  680583894916304897
getting data for:  680497766108381184
getting data for:  680494726643068929
getting data for:  680473011644985345
getting data for:  680440374763077632
getting data for:  680221482581123072
getting data

getting data for:  674644256330530816
getting data for:  674638615994089473
getting data for:  674632714662858753
getting data for:  674606911342424069
getting data for:  674468880899788800
getting data for:  674447403907457024
getting data for:  674436901579923456
getting data for:  674422304705744896
getting data for:  674416750885273600
getting data for:  674410619106390016
getting data for:  674394782723014656
getting data for:  674372068062928900
getting data for:  674330906434379776
getting data for:  674318007229923329
getting data for:  674307341513269249
getting data for:  674291837063053312
getting data for:  674271431610523648
getting data for:  674269164442398721
getting data for:  674265582246694913
getting data for:  674262580978937856
getting data for:  674255168825880576
getting data for:  674082852460433408
getting data for:  674075285688614912
getting data for:  674063288070742018
getting data for:  674053186244734976
getting data for:  674051556661161984
getting data

getting data for:  670420569653809152
getting data for:  670417414769758208
getting data for:  670411370698022913
getting data for:  670408998013820928
getting data for:  670403879788544000
getting data for:  670385711116361728
getting data for:  670374371102445568
getting data for:  670361874861563904
getting data for:  670338931251150849
getting data for:  670319130621435904
getting data for:  670303360680108032
getting data for:  670290420111441920
getting data for:  670093938074779648
getting data for:  670086499208155136
getting data for:  670079681849372674
getting data for:  670073503555706880
getting data for:  670069087419133954
getting data for:  670061506722140161
getting data for:  670055038660800512
getting data for:  670046952931721218
getting data for:  670040295598354432
getting data for:  670037189829525505
getting data for:  670003130994700288
getting data for:  669993076832759809
getting data for:  669972011175813120
getting data for:  669970042633789440
getting data

getting data for:  666293911632134144
getting data for:  666287406224695296
getting data for:  666273097616637952
getting data for:  666268910803644416
getting data for:  666104133288665088
getting data for:  666102155909144576
getting data for:  666099513787052032
getting data for:  666094000022159362
getting data for:  666082916733198337
getting data for:  666073100786774016
getting data for:  666071193221509120
getting data for:  666063827256086533
getting data for:  666058600524156928
getting data for:  666057090499244032
getting data for:  666055525042405380
getting data for:  666051853826850816
getting data for:  666050758794694657
getting data for:  666049248165822465
getting data for:  666044226329800704
getting data for:  666033412701032449
getting data for:  666029285002620928
getting data for:  666020888022790149


### Test
Reading the stored json data into pandas dataframe

In [104]:
len(exception_list)

21

21 twitter ids from the archive do not have existing data from

In [90]:
status = []
with open('tweet-json.txt','r') as file:
    for line in file:
        status.append(json.loads(line))


In [93]:
dicts = []
for string in status[0]:
    dicts.append(json.loads(string))

In [95]:
type(dicts[0])

dict

In [97]:
tweet_df = pd.DataFrame(dicts)
tweet_df.head()

contributors coordinates                      created_at display_text_range  \
0         None        None  Mon Nov 16 00:24:50 +0000 2015           [0, 120]   
1         None        None  Mon Nov 16 00:04:52 +0000 2015           [0, 137]   
2         None        None  Sun Nov 15 23:21:54 +0000 2015           [0, 130]   
3         None        None  Sun Nov 15 23:05:30 +0000 2015           [0, 139]   
4         None        None  Sun Nov 15 22:32:08 +0000 2015           [0, 131]   

                                            entities  \
0  {'hashtags': [], 'symbols': [], 'user_mentions...   
1  {'hashtags': [], 'symbols': [], 'user_mentions...   
2  {'hashtags': [], 'symbols': [], 'user_mentions...   
3  {'hashtags': [], 'symbols': [], 'user_mentions...   
4  {'hashtags': [], 'symbols': [], 'user_mentions...   

                                   extended_entities  favorite_count  \
0  {'media': [{'id': 666049244999131136, 'id_str'...             104   
1  {'media': [{'id': 666044217047650304, 'id_str'...             288   
2  {'media': [{'id': 666033409081393153, 'id_str'...             120   
3  {'media': [{'id': 666029276303482880, 'id_str'...             124   
4  {'media': [{'id': 666020881337073664, 'id_str'...            2503   

   favorited                                          full_text   geo  \
0      False  Here we have a 1949 1st generation vulpix. Enj...  None   
1      False  This is a purebred Piers Morgan. Loves to Netf...  None   
2      False  Here is a very happy pup. Big fan of well-main...  None   
3      False  This is a western brown Mitsubishi terrier. Up...  None   
4      False  Here we have a Japanese Irish Setter. Lost eye...  None   

                         ...                          quoted_status  \
0                        ...                                    NaN   
1                        ...                                    NaN   
2                        ...                                    NaN   
3                        ...                                    NaN   
4                        ...                                    NaN   

  quoted_status_id quoted_status_id_str  quoted_status_permalink  \
0              NaN                  NaN                      NaN   
1              NaN                  NaN                      NaN   
2              NaN                  NaN                      NaN   
3              NaN                  NaN                      NaN   
4              NaN                  NaN                      NaN   

  retweet_count  retweeted retweeted_status  \
0            41      False              NaN   
1           135      False              NaN   
2            43      False              NaN   
3            46      False              NaN   
4           493      False              NaN   

                                              source truncated  \
0  <a href="http://twitter.com/download/iphone" r...     False   
1  <a href="http://twitter.com/download/iphone" r...     False   
2  <a href="http://twitter.com/download/iphone" r...     False   
3  <a href="http://twitter.com/download/iphone" r...     False   
4  <a href="http://twitter.com/download/iphone" r...     False   

                                                user  
0  {'id': 4196983835, 'id_str': '4196983835', 'na...  
1  {'id': 4196983835, 'id_str': '4196983835', 'na...  
2  {'id': 4196983835, 'id_str': '4196983835', 'na...  
3  {'id': 4196983835, 'id_str': '4196983835', 'na...  
4  {'id': 4196983835, 'id_str': '4196983835', 'na...  

[5 rows x 32 columns]

## Gather the image predictions dataset using requests
download the dataset from https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv

In [23]:
import requests

In [25]:
response = requests.get('https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv')

In [28]:
response

<Response [200]>

In [29]:
with open('image-predictions.tsv','w') as file:
    file.write(response.text)

In [30]:
image_pred_df = pd.read_csv('image-predictions.tsv',sep = '\t')
image_pred_df.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

## Gathered Data
- tweet_df -contains tweet data from querying each tweet id from the archive dataset
- tweet_archive_df -contains all the tweets of we rate dogs, that contain some rating for a dog
- imge_pred_df -contains a dataframe of predictions of the breed of each dog in the tweet archive dataframe

# Assess

In [98]:
tweet_df.head()

contributors coordinates                      created_at display_text_range  \
0         None        None  Mon Nov 16 00:24:50 +0000 2015           [0, 120]   
1         None        None  Mon Nov 16 00:04:52 +0000 2015           [0, 137]   
2         None        None  Sun Nov 15 23:21:54 +0000 2015           [0, 130]   
3         None        None  Sun Nov 15 23:05:30 +0000 2015           [0, 139]   
4         None        None  Sun Nov 15 22:32:08 +0000 2015           [0, 131]   

                                            entities  \
0  {'hashtags': [], 'symbols': [], 'user_mentions...   
1  {'hashtags': [], 'symbols': [], 'user_mentions...   
2  {'hashtags': [], 'symbols': [], 'user_mentions...   
3  {'hashtags': [], 'symbols': [], 'user_mentions...   
4  {'hashtags': [], 'symbols': [], 'user_mentions...   

                                   extended_entities  favorite_count  \
0  {'media': [{'id': 666049244999131136, 'id_str'...             104   
1  {'media': [{'id': 666044217047650304, 'id_str'...             288   
2  {'media': [{'id': 666033409081393153, 'id_str'...             120   
3  {'media': [{'id': 666029276303482880, 'id_str'...             124   
4  {'media': [{'id': 666020881337073664, 'id_str'...            2503   

   favorited                                          full_text   geo  \
0      False  Here we have a 1949 1st generation vulpix. Enj...  None   
1      False  This is a purebred Piers Morgan. Loves to Netf...  None   
2      False  Here is a very happy pup. Big fan of well-main...  None   
3      False  This is a western brown Mitsubishi terrier. Up...  None   
4      False  Here we have a Japanese Irish Setter. Lost eye...  None   

                         ...                          quoted_status  \
0                        ...                                    NaN   
1                        ...                                    NaN   
2                        ...                                    NaN   
3                        ...                                    NaN   
4                        ...                                    NaN   

  quoted_status_id quoted_status_id_str  quoted_status_permalink  \
0              NaN                  NaN                      NaN   
1              NaN                  NaN                      NaN   
2              NaN                  NaN                      NaN   
3              NaN                  NaN                      NaN   
4              NaN                  NaN                      NaN   

  retweet_count  retweeted retweeted_status  \
0            41      False              NaN   
1           135      False              NaN   
2            43      False              NaN   
3            46      False              NaN   
4           493      False              NaN   

                                              source truncated  \
0  <a href="http://twitter.com/download/iphone" r...     False   
1  <a href="http://twitter.com/download/iphone" r...     False   
2  <a href="http://twitter.com/download/iphone" r...     False   
3  <a href="http://twitter.com/download/iphone" r...     False   
4  <a href="http://twitter.com/download/iphone" r...     False   

                                                user  
0  {'id': 4196983835, 'id_str': '4196983835', 'na...  
1  {'id': 4196983835, 'id_str': '4196983835', 'na...  
2  {'id': 4196983835, 'id_str': '4196983835', 'na...  
3  {'id': 4196983835, 'id_str': '4196983835', 'na...  
4  {'id': 4196983835, 'id_str': '4196983835', 'na...  

[5 rows x 32 columns]

In [103]:
tweet_df.columns

Index(['contributors', 'coordinates', 'created_at', 'display_text_range',
       'entities', 'extended_entities', 'favorite_count', 'favorited',
       'full_text', 'geo', 'id', 'id_str', 'in_reply_to_screen_name',
       'in_reply_to_status_id', 'in_reply_to_status_id_str',
       'in_reply_to_user_id', 'in_reply_to_user_id_str', 'is_quote_status',
       'lang', 'place', 'possibly_sensitive', 'possibly_sensitive_appealable',
       'quoted_status', 'quoted_status_id', 'quoted_status_id_str',
       'quoted_status_permalink', 'retweet_count', 'retweeted',
       'retweeted_status', 'source', 'truncated', 'user'],
      dtype='object')

In [118]:
tweet_archive_df

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
19    888202515573088257                    NaN                  NaN   
20    888078434458587136                    NaN                  NaN   
21    887705289381826560                    NaN                  NaN   
22    887517139158093824                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
26    886983233522544640                    NaN                  NaN   
27    886736880519319552                    NaN                  NaN   
28    886680336477933568                    NaN                  NaN   
29    886366144734445568                    NaN                  NaN   
...                  ...                    ...                  ...   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2328  666396247373291520                    NaN                  NaN   
2329  666373753744588802                    NaN                  NaN   
2330  666362758909284353                    NaN                  NaN   
2331  666353288456101888                    NaN                  NaN   
2332  666345417576210432                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2336  666273097616637952                    NaN                  NaN   
2337  666268910803644416                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
2339  666102155909144576                    NaN                  NaN   
2340  666099513787052032                    NaN                  NaN   
2341  666094000022159362                    NaN                  NaN   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2349  666051853826850816                    NaN         

In [119]:
tweet_archive_df[['doggo','floofer','pupper','puppo']].describe()

doggo floofer pupper puppo
count   2356    2356   2356  2356
unique     2       2      2     2
top     None    None   None  None
freq    2259    2346   2099  2326

In [124]:
tweet_archive_df.doggo.value_counts()

None     2259
doggo      97
Name: doggo, dtype: int64

In [125]:
tweet_archive_df.pupper.value_counts()

None      2099
pupper     257
Name: pupper, dtype: int64

In [126]:
tweet_archive_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [128]:
tweet_archive_df.rating_denominator.value_counts()

10     2333
11        3
50        3
80        2
20        2
2         1
16        1
40        1
70        1
15        1
90        1
110       1
120       1
130       1
150       1
170       1
7         1
0         1
Name: rating_denominator, dtype: int64

In [129]:
tweet_archive_df.rating_numerator.value_counts()

12      558
11      464
10      461
13      351
9       158
8       102
7        55
14       54
5        37
6        32
3        19
4        17
1         9
2         9
420       2
0         2
15        2
75        2
80        1
20        1
24        1
26        1
44        1
50        1
60        1
165       1
84        1
88        1
144       1
182       1
143       1
666       1
960       1
1776      1
17        1
27        1
45        1
99        1
121       1
204       1
Name: rating_numerator, dtype: int64

In [137]:
tweet_archive_df['text'].value_counts()

This is Lola. She realized mid hug that she's not ready for a committed relationship with a teddy bear. 9/10 https://t.co/pVebzwRioD                                      1
This is Duke. He is not a fan of the pupporazzi. 12/10 https://t.co/SgpBVYIL18                                                                                            1
This is Eleanor. She winks like she knows many things that you don't. 12/10 https://t.co/bxGwkJa2kE                                                                       1
This is Sparky. That's his pancake now. He will raise it as his own. 10/10 https://t.co/96tMaWyoWt                                                                        1
Meet Larry. He doesn't know how to shoe. 9/10 damn it Larry https://t.co/jMki5GOV3y                                                                                       1
We only rate dogs. Pls stop sending non-canines like this Bulgarian Eyeless Porch Bear. This is unacceptable... 9/10 https://t.co/2yctWAUZ3Z

In [133]:
image_pred_df

tweet_id                                            jpg_url  \
0     666020888022790149    https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928    https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449    https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704    https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465    https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
5     666050758794694657    https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg   
6     666051853826850816    https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
7     666055525042405380    https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg   
8     666057090499244032    https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
9     666058600524156928    https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg   
10    666063827256086533    https://pbs.twimg.com/media/CT5Vg_wXIAAXfnj.jpg   
11    666071193221509120    https://pbs.twimg.com/media/CT5cN_3WEAAlOoZ.jpg   
12    666073100786774016    https://pbs.twimg.com/media/CT5d9DZXAAALcwe.jpg   
13    666082916733198337    https://pbs.twimg.com/media/CT5m4VGWEAAtKc8.jpg   
14    666094000022159362    https://pbs.twimg.com/media/CT5w9gUW4AAsBNN.jpg   
15    666099513787052032    https://pbs.twimg.com/media/CT51-JJUEAA6hV8.jpg   
16    666102155909144576    https://pbs.twimg.com/media/CT54YGiWUAEZnoK.jpg   
17    666104133288665088    https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg   
18    666268910803644416    https://pbs.twimg.com/media/CT8QCd1WEAADXws.jpg   
19    666273097616637952    https://pbs.twimg.com/media/CT8T1mtUwAA3aqm.jpg   
20    666287406224695296    https://pbs.twimg.com/media/CT8g3BpUEAAuFjg.jpg   
21    666293911632134144    https://pbs.twimg.com/media/CT8mx7KW4AEQu8N.jpg   
22    666337882303524864    https://pbs.twimg.com/media/CT9OwFIWEAMuRje.jpg   
23    666345417576210432    https://pbs.twimg.com/media/CT9Vn7PWoAA_ZCM.jpg   
24    666353288456101888    https://pbs.twimg.com/media/CT9cx0tUEAAhNN_.jpg   
25    666362758909284353    https://pbs.twimg.com/media/CT9lXGsUcAAyUFt.jpg   
26    666373753744588802    https://pbs.twimg.com/media/CT9vZEYWUAAlZ05.jpg   
27    666396247373291520    https://pbs.twimg.com/media/CT-D2ZHWIAA3gK1.jpg   
28    666407126856765440    https://pbs.twimg.com/media/CT-NvwmW4AAugGZ.jpg   
29    666411507551481857    https://pbs.twimg.com/media/CT-RugiWIAELEaq.jpg   
...                  ...                                                ...   
2045  886366144734445568    https://pbs.twimg.com/media/DE0BTnQUwAApKEH.jpg   
2046  886680336477933568    https://pbs.twimg.com/media/DE4fEDzWAAAyHMM.jpg   
2047  886736880519319552    https://pbs.twimg.com/media/DE5Se8FXcAAJFx4.jpg   
2048  886983233522544640    https://pbs.twimg.com/media/DE8yicJW0AAAvBJ.jpg   
2049  887101392804085760    https://pbs.twimg.com/media/DE-eAq6UwAA-jaE.jpg   
2050  887343217045368832  https://pbs.twimg.com/ext_tw_video_thumb/88734...   
2051  887473957103951883    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2052  887517139158093824  https://pbs.twimg.com/ext_tw_video_thumb/88751...   
2053  887705289381826560    https://pbs.twimg.com/media/DFHDQBbXgAEqY7t.jpg   
2054  888078434458587136    https://pbs.twimg.com/media/DFMWn56WsAAkA7B.jpg   
2055  888202515573088257    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2056  888554962724278272    https://pbs.twimg.com/media/DFTH_O-UQAACu20.jpg   
2057  888804989199671297    https://pbs.twimg.com/media/DFWra-3VYAA2piG.jpg   
2058  888917238123831296    https://pbs.twimg.com/media/DFYRgsOUQAARGhO.jpg   
2059  889278841981685760  https://pbs.twimg.com/ext_tw_video_thumb/88927...   
2060  889531135344209921    https://pbs.twimg.com/media/DFg_2PVW0AEHN3p.jpg   
2061  889638837579907072    https://pbs.twimg.com/media/DFihzFfXsAYGDPR.jpg   
2062  889665388333682689    https://pbs.twimg.com/media/DFi579UWsAAatzw.jpg   
2063  889880896479866881    https://pbs.twimg.com/media/DFl99B1WsA

In [132]:
image_pred_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


## Issues
### Tidyness
1. retweet and favorite count of tweet_df not joined with tweet_archive_df

##### tweet_archive_df
1. stage of dog is 1 column but is divided into 4 columns: doggo, floofer, pupper, puppo

### Quality
##### tweet_archive_df
1. several dogs do not have any dog stage assigned.
2. Dog stage Floof is named floofer
3. timestamp column is of object type and not datetime
4. missing values in dataset are replaced by 'None' text. e.g.doggo, floofer, pupper, puppo
5. retweeted status timestamp is of object type and not datetime
6. rating_denominator values have values other than 10
7. rating_numerator values have some values greater than 100
8. There are 181 retweeted tweets, and not original ratings. found from retweeted_to column
9. Some of the rating numerators are wrongly extracted from text. e.g.row:314, numerator is extracted as 960, but actually 13
10. Some tweets have more than 1 dog stage

##### image_pred_df
1. There are 2075 tweets that contain images.

# Clean
## tidyness issue 1: merging datasets
### Define
Merge the datasets using the pandas merge method. use inner join, so only those rows that are common to both the tables remain.

<br> Since only 21 rows do not have corresponding data in the tweet_df out of a total of 2356 entries, this is not a significant omission.
### Code

##### Merging tweet datasets
Bringing retweet and favorite counts into the tweet archive dataframe

In [107]:
tweet_archive_clean_df = tweet_archive_df.copy()


In [115]:
tweet_archive_clean_df = tweet_archive_clean_df.merge(tweet_df[['id','retweet_count','favorite_count']],how='inner',left_on='tweet_id',right_on='id',
              right_index=False)

### Test

In [116]:
tweet_archive_clean_df.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  \
0                  10   Phineas  None    None   None  None   
1                  10     Tilly  None    None   None  None   
2                  10    Archie  None    None   None  None   
3                  10     Darla  None    None   None  None   
4                  10  Franklin  None    None   None  None   

                   id  retweet_count  favorite_count  
0  892420643555336193           8178           37517  
1  892177421306343426           6051           32258  
2  891815181378084864           4003           24298  
3  891689557279858688           8318           40849  
4  891327558926688256           9025           39062

In [117]:
tweet_archive_clean_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2345 entries, 0 to 2344
Data columns (total 20 columns):
tweet_id                      2345 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2345 non-null object
source                        2345 non-null object
text                          2345 non-null object
retweeted_status_id           166 non-null float64
retweeted_status_user_id      166 non-null float64
retweeted_status_timestamp    166 non-null object
expanded_urls                 2286 non-null object
rating_numerator              2345 non-null int64
rating_denominator            2345 non-null int64
name                          2345 non-null object
doggo                         2345 non-null object
floofer                       2345 non-null object
pupper                        2345 non-null object
puppo                         2345 non-null object
id                            23

## Tidyness issue 2: Stage of dog is distributed into 4 columns
## Quality issue 4: Handle 'None' values in dog stage columns

### Define
- combine the 4 columns, keep only non none values

### Code

In [179]:
c = tweet_archive_clean_df[['doggo','floofer','pupper','puppo']].replace('None','')
c['stage'] = c.doggo + c.floofer + c.pupper + c.puppo
c.stage.value_counts()

                1969
pupper           242
doggo             82
puppo             29
doggopupper       12
floofer            9
doggopuppo         1
doggofloofer       1
Name: stage, dtype: int64

In [186]:
tweet_archive_clean_df[c.stage == 'floofer'].text.values

array(['Meet Grizzwald. He may be the floofiest floofer I ever did see. Lost eyes saving a schoolbus from a volcano erpuption. 13/10 heroic as h*ck https://t.co/rf661IFEYP',
       'This is Doc. He takes time out of every day to worship our plant overlords. 12/10 quite the floofer https://t.co/azMneS6Ly5',
       "Atlas rolled around in some chalk and now he's a magical rainbow floofer. 13/10 please never take a bath https://t.co/nzqTNw0744",
       "This is Blu. He's a wild bush Floofer. I wish anything made me as happy as bushes make Blu. 12/10 would frolic with https://t.co/HHUAnBb6QB",
       "Here's a golden floofer helping with the groceries. Bed got in way. Still 11/10 helpful af (vid by @categoen) https://t.co/6ZRoZUWFmd",
       "Just wanted to share this super rare Rainbow Floofer in case you guys haven't seen it yet. 13/10 colorful af https://t.co/CaG9MzD3WT",
       "This is Moose. He's a Polynesian Floofer. Dapper af. 10/10 would pet diligently https://t.co/mVfqRdppTL",
  

In [ ]:
tweet_archive_clean_df['stage'] = c.stage

### Test

In [187]:
tweet_archive_clean_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2345 entries, 0 to 2344
Data columns (total 21 columns):
tweet_id                      2345 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2345 non-null object
source                        2345 non-null object
text                          2345 non-null object
retweeted_status_id           166 non-null float64
retweeted_status_user_id      166 non-null float64
retweeted_status_timestamp    166 non-null object
expanded_urls                 2286 non-null object
rating_numerator              2345 non-null int64
rating_denominator            2345 non-null int64
name                          2345 non-null object
doggo                         2345 non-null object
floofer                       2345 non-null object
pupper                        2345 non-null object
puppo                         2345 non-null object
id                            23

## Quality issue 1: Handle missing dog stage data
### Define
- replace the missing values with 'not_mentioned'

### Code

In [191]:
tweet_archive_clean_df.stage.replace('','not_mentioned',inplace=True)

### Test

In [192]:
tweet_archive_clean_df.stage.value_counts()

not_mentioned    1969
pupper            242
doggo              82
puppo              29
doggopupper        12
floofer             9
doggopuppo          1
doggofloofer        1
Name: stage, dtype: int64

## Quality issue 2: dog stage floof is mentioned as floofer
### Define
- replace every occurance of floofer with floof in dog stage column

### Code

In [195]:
tweet_archive_clean_df.stage.replace('floofer','floof',inplace=True)

### Test

In [196]:
tweet_archive_clean_df.stage.value_counts()

not_mentioned    1969
pupper            242
doggo              82
puppo              29
doggopupper        12
floof               9
doggopuppo          1
doggofloofer        1
Name: stage, dtype: int64

## Quality issue 3: Time stamp is of type object
### Define
- Convert time stamp column to date format

### Code

In [203]:
tweet_archive_clean_df.timestamp = pd.to_datetime(tweet_archive_clean_df.timestamp)

### Test

In [204]:
tweet_archive_clean_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2345 entries, 0 to 2344
Data columns (total 21 columns):
tweet_id                      2345 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2345 non-null datetime64[ns]
source                        2345 non-null object
text                          2345 non-null object
retweeted_status_id           166 non-null float64
retweeted_status_user_id      166 non-null float64
retweeted_status_timestamp    166 non-null object
expanded_urls                 2286 non-null object
rating_numerator              2345 non-null int64
rating_denominator            2345 non-null int64
name                          2345 non-null object
doggo                         2345 non-null object
floofer                       2345 non-null object
pupper                        2345 non-null object
puppo                         2345 non-null object
id                      

## Issues left


### Quality
##### tweet_archive_df
5. retweeted status timestamp is of object type and not datetime
6. rating_denominator values have values other than 10
7. rating_numerator values have some values greater than 100
8. There are 181 retweeted tweets, and not original ratings. found from retweeted_to column
9. Some of the rating numerators are wrongly extracted from text. e.g.row:314, numerator is extracted as 960, but actually 13
10. Some tweets have more than 1 dog stage

##### image_pred_df
1. There are 2075 tweets that contain images.

## Quality issue 5: retweeted Time stamp is of type object
### Define
- Convert retweeted time stamp column to date format

### Code

In [205]:
tweet_archive_clean_df.retweeted_status_timestamp = pd.to_datetime(tweet_archive_clean_df.retweeted_status_timestamp)

### Test

In [206]:
tweet_archive_clean_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2345 entries, 0 to 2344
Data columns (total 21 columns):
tweet_id                      2345 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2345 non-null datetime64[ns]
source                        2345 non-null object
text                          2345 non-null object
retweeted_status_id           166 non-null float64
retweeted_status_user_id      166 non-null float64
retweeted_status_timestamp    166 non-null datetime64[ns]
expanded_urls                 2286 non-null object
rating_numerator              2345 non-null int64
rating_denominator            2345 non-null int64
name                          2345 non-null object
doggo                         2345 non-null object
floofer                       2345 non-null object
pupper                        2345 non-null object
puppo                         2345 non-null object
id              

## Quality issue 6: denominator values of rating have varied values other than 10
### Define
- check the genuineness of the ratings
- extract the correct denominator and replace in the necessary column

### Code

In [214]:
c = tweet_archive_clean_df.copy()
#c.rating_denominator.value_counts()
c[c.rating_denominator!=10].text.values[0]

"@jonnysun @Lin_Manuel ok jomny I know you're excited but 960/00 isn't a valid rating, 13/10 is tho"

In [215]:
c[c.rating_denominator!=10].text.values[1]

'@docmisterio account started on 11/15/15'

In [216]:
c[c.rating_denominator!=10].text.values[2]

'The floofs have been released I repeat the floofs have been released. 84/70 https://t.co/NIYC820tmd'

In [217]:
c[c.rating_denominator!=10].text.values[3]

'Meet Sam. She smiles 24/7 &amp; secretly aspires to be a reindeer. \nKeep Sam smiling by clicking and sharing this link:\nhttps://t.co/98tB8y7y7t https://t.co/LouL5vdvxx'

## Iterate on Assess Step
As we see there are some corrections that are required in the denominator of the rating, we are iterating on our assess step once again.

In [222]:
print(c[c.rating_denominator >=10].text.values)

["This is Phineas. He's a mystical boy. Only ever appears in the hole of a donut. 13/10 https://t.co/MgUWQ76dJU"
 "This is Tilly. She's just checking pup on you. Hopes you're doing ok. If not, she's available for pats, snugs, boops, the whole bit. 13/10 https://t.co/0Xxu71qeIV"
 'This is Archie. He is a rare Norwegian Pouncing Corgo. Lives in the tall grass. You never know when one may strike. 12/10 https://t.co/wUnZnhtVJB'
 ...
 'Here we have a Japanese Irish Setter. Lost eye in Vietnam (?). Big fan of relaxing on stair. 8/10 would pet https://t.co/BLDqew2Ijj'
 'Here we have a Japanese Irish Setter. Lost eye in Vietnam (?). Big fan of relaxing on stair. 8/10 would pet https://t.co/BLDqew2Ijj'
 'Here we have a Japanese Irish Setter. Lost eye in Vietnam (?). Big fan of relaxing on stair. 8/10 would pet https://t.co/BLDqew2Ijj']


In [221]:
print(c[c.rating_denominator <10].text.values)

["@jonnysun @Lin_Manuel ok jomny I know you're excited but 960/00 isn't a valid rating, 13/10 is tho"
 'Meet Sam. She smiles 24/7 &amp; secretly aspires to be a reindeer. \nKeep Sam smiling by clicking and sharing this link:\nhttps://t.co/98tB8y7y7t https://t.co/LouL5vdvxx'
 'This is an Albanian 3 1/2 legged  Episcopalian. Loves well-polished hardwood flooring. Penis on the collar. 9/10 https://t.co/d9NcXFKwLv']


Assessing from the above, we see that ratings can have more or less than 10 as denominator

In [227]:
c[c.rating_denominator ==0].text.values

array(["@jonnysun @Lin_Manuel ok jomny I know you're excited but 960/00 isn't a valid rating, 13/10 is tho"],
      dtype=object)

## More issues
#### tweet_archive_df
- rating_denominator is actually 10 where it shows 0